libbeat/common/mapstr.go

```diff
@@ -257,6 +257,25 @@ func (m MapStr) Format(f fmt.State, c rune) {
 	io.WriteString(f, debugM.String())
 }
 
+// Expand does the reverse of Flatten - expands dotted test and return an expanded MapStr.
+//
+// Example:
+//   "hello.world": "test"
+//
+// This is converted to:
+//   "hello": MapStr{"world": "test" }
+//
+func (m MapStr) Expand() (MapStr, error) {
+	e := MapStr{}
+	f := m.Flatten()
+	for k, v := range f {
+		if _, err := e.Put(k, v); err != nil {
+			return e, err
+		}
+	}
+	return e, nil
+}
```

Comment:
Yay, more methods on common.MapStr. MapStr interface is somewhat inconsistent, but as the Expand interface looks to be consistent with 'Flatten', I think the interface is ok as is. btw. nice trick about using 'Put' :)

As you may have noticed during testing, this approach is non-deterministic. Whenever go creates a map, it generates a random 'seed' value to be used when hashing keys. Thusly keys are ordered randomly. This can actually lead to conflicts when expanding, e.g. if we have `a.b=1` and `a.b.c=2`. Is `a.b` now a primitive or an object. In order to establish some determinism you need to create a sorted array of keys first (sorting based on length would be good enough).

Not sure where, but I think somewhere we solved the conflict by renaming `a.b=1` to `a.b.value=1`